### ISSUE TITLE
Mail Queue Problem Illuminate\Mail with Lumen

### ISSUE THREAD

**Issue by rlacerda83** [laravel/framework#10083]
I got fatal error: Call to a member function push() when run:

```php
Mail::queue('email.blank', ['html' => $html], function($msg) {
            $msg->to(['email@gmail.com']);
            $msg->from(['email@gmail.com']);
            $msg->subject('Lumen Service Mail');
        });
```

Its work if change file: Illuminate\Mail\Mailer.php

Add
`use \Illuminate\Support\Facades\Queue as Queue;`

And change line 205:
from: `return $this->queue->push('mailer@handleQueuedMessage', compact('view', 'data', 'callback'), $queue);`
to `return Queue::push('mailer@handleQueuedMessage', compact('view', 'data', 'callback'), $queue);`

this work for me!

**Comment by GrahamCampbell** [comment]
That's NOT the correct solution.

**Comment by GrahamCampbell** [comment]
All you need to do is make sure you've used the queue before you use mail. That's just due to the way Lumen doesn't even bother to bind stuff unless you try to use it, to save time. If you don't like that behaviour, I'd recommend just using Laravel.

https://github.com/laravel/framework/blob/5.1/src/Illuminate/Mail/MailServiceProvider.php#L77

**Comment by rlacerda83** [comment]
Thank you for your response.

Im using the queue only to send emails. Do you have any suggestion to start the queue before send email? 

-----Mensagem Original-----
De: "Graham Campbell" <notifications@github.com>
Enviada em: ‎30/‎08/‎2015 04:54
Para: "laravel/framework" <framework@noreply.github.com>
Cc: "Rodrigo Lacerda" <r.lacerda83@gmail.com>
Assunto: Re: [framework] Mail Queue Problem Illuminate\Mail with Lumen(#10083)

Closed #10083.

**Comment by friedolinfoerder** [comment]
I have the same issue and I would also like to know how to start the queue before sending email.
This would also add a valid solution for this issue: https://github.com/laravel/lumen-framework/issues/77

**Comment by lelu85** [comment]
Like graham said this is not the right solution. You can check the implementation of the Illuminate\Mail\Mailer class and you will see it has a setQueue method so by doing:

    public function __construct()
    {
        $this->mailer = app('mailer');
        $this->mailer->setQueue(app('queue')->connection('redis'));
    }

It is done like that. You can even inject the Mailer or the QueueContract with bindings, however you want.

**Comment by ishannz** [comment]
Create a new job - use the follwoing sample code
<pre>
namespace App\Jobs;
use App\Jobs\Job;

class SendEmail extends Job
{
    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        //
    }
} 
</pre>

and in your controller 

include the SendMail under the namepsace;
<pre>
use App\Jobs\SendEmail;
</pre>
mail sending method
<pre>
public function mailSend()
{
        $data = array(
            'title' => 'test title',
            'description' => 'test description',
        );
        Queue::push(new SendEmail);

        Mail::queue('email.mytemplate', $data, function ($message) {
            $message->from('ishan@testmail.com', 'Ishan test mail server');
            $message->to('ishanmail@testmail.com')->subject('test subject');
        });
}
</pre>

**Comment by chimit** [comment]
@GrahamCampbell if it's an intended behavior what is the right way to test mailables?

**Comment by chimit** [comment]
Looks like this problem was fixed in Lumen 5.5.

**Comment by hakuno** [comment]
@chimit The queue isn't being `bound`ed in the mail service provider. So the `setQueue` isn't called properly.

I found a solution for 5.4's.

Look at here: https://github.com/laravel/lumen-framework/issues/590#issuecomment-316363236

**Comment by simonhamp** [comment]
FYI @GrahamCampbell this is a bug. IMHO, all Illuminate elements are should be made to work out of the box with Laravel and Lumen even if they are not explicitly included.

The best fix I've found for this (in Lumen 5.5) is to do the following:

In `bootstrap/app.php`, make sure the `MailServiceProvider` is listed **before** your `AppServiceProvider`, e.g.:
```php
...
$app->register(Illuminate\Mail\MailServiceProvider::class);
$app->register(App\Providers\AppServiceProvider::class);
...
```

Then along the lines of @soderluk's [comment on this thread](https://github.com/laravel/lumen-framework/issues/590#issuecomment-316363236), in your `AppServiceProvider` add the following in your `register` method:
```php
app('mailer')->setQueue(app('queue'));
```
